services/wfm/src/main/java/org/openkilda/wfm/topology/switchmanager/service/impl/CommandBuilderImpl.java

```diff
@@ -136,21 +138,44 @@ public CommandBuilderImpl(PersistenceManager persistenceManager, FlowResourcesCo
                 .collect(Collectors.toList());
     }
 
-    private RemoveFlow buildRemoveFlowWithoutMeterFromFlowEntry(SwitchId switchId, FlowEntry entry) {
+    @VisibleForTesting
+    RemoveFlow buildRemoveFlowWithoutMeterFromFlowEntry(SwitchId switchId, FlowEntry entry) {
         Optional<FlowMatchField> entryMatch = Optional.ofNullable(entry.getMatch());
+        Optional<FlowInstructions> instructions = Optional.ofNullable(entry.getInstructions());
+        Optional<FlowApplyActions> applyActions = instructions.map(FlowInstructions::getApplyActions);
 
         Integer inPort = entryMatch.map(FlowMatchField::getInPort).map(Integer::valueOf).orElse(null);
+
+        FlowEncapsulationType encapsulationType = FlowEncapsulationType.TRANSIT_VLAN;
+        Integer encapsulationId = null;
         Integer vlan = entryMatch.map(FlowMatchField::getVlanVid).map(Integer::valueOf).orElse(null);
+        if (vlan != null) {
+            encapsulationId = vlan;
+        } else {
+            Integer tunnelId = entryMatch.map(FlowMatchField::getTunnelId).map(Integer::valueOf).orElse(null);
+            if (tunnelId == null) {
+                tunnelId = applyActions.map(FlowApplyActions::getPushVxlan).map(Integer::valueOf).orElse(null);
+            }
+
+            if (tunnelId != null) {
+                encapsulationId = tunnelId;
+                encapsulationType = FlowEncapsulationType.VXLAN;
+            }
+        }
 
         Optional<FlowApplyActions> actions = Optional.ofNullable(entry.getInstructions())
                 .map(FlowInstructions::getApplyActions);
 
-        Integer outPort = actions.map(FlowApplyActions::getFlowOutput).map(Integer::valueOf).orElse(null);
+        Integer outPort = actions
+                .map(FlowApplyActions::getFlowOutput)
+                .filter(NumberUtils::isNumber)
```

Comment:
Why `OFPort.IN_PORT` is not a number?

Reply:
```
public static final OFPort IN_PORT = new OFPort.NamedPort(-8, "in_port");
```
More research is needed to represent this field by number. We do not know how the system will behave if negative ports return.

Reply:
All OF the ports represents with 32-bit number (unsigned  - because they have unsigned types)... There is a range for special ports (in_port is one of them). So - one more time - why we have port represented as not a number? Do we really need to pass it as "out-port" into flow-mod message?

This whole ... piece of code with "unpredictable" flow-mod(delete) request is going to be removed soon(as I hope)... so... I remove my -1.